lib/propolis/src/hw/virtio/softnpu.rs

```diff
@@ -0,0 +1,1061 @@
+        Arc::new(PciVirtioSoftNpuPort {
+            mac,
+            data_handles,
+            pipeline,
+            log,
+            virtio_state: virtio.clone(),
+        })
+    }
+
+    fn handle_guest_virtio_request(&self, vq: &Arc<VirtQueue>) {
+        if vq.id == 0 {
+            return self.handle_q0_req(vq);
+        }
+
+        let mem = match self.virtio_state.pci_state.acc_mem.access() {
+            Some(mem) => mem,
+            None => {
+                warn!(self.log, "failed to access virtio memory");
+                return;
+            }
+        };
+        let mut chain = Chain::with_capacity(1);
+        match vq.pop_avail(&mut chain, &mem) {
+            Some(val) => val as usize,
+            None => return,
+        };
+
+        // only vq.push_used if we actually read something
+        let mut push_used = false;
+
+        // read as many Ethernet frames from the guest as we can
+        loop {
+            let mut virtio_bytes = [0u8; 10];
+            // read in virtio mystery bytes
+            let n = read_buf(&mem, &mut chain, &mut virtio_bytes);
+            if n != 10 {
+                if n > 0 {
+                    push_used = true;
+                }
+                warn!(self.log, "failed to read virtio mystery bytes ({})", n);
+                //break;
+            }
+
+            let mut frame = [0u8; MTU];
+            // read in Ethernet header
+            let n = read_buf(&mem, &mut chain, &mut frame);
+            if n == 0 {
+                break;
+            }
+            push_used = true;
+
+            let pkt = packet_in::new(&frame[..n]);
+
+            let mut pipeline = match self.pipeline.lock() {
+                Ok(pipe) => pipe,
+                Err(e) => {
+                    error!(self.log, "failed to lock pipeline: {}", e);
+                    break;
+                }
+            };
+            let pl: &mut Box<dyn Pipeline> = match &mut *pipeline {
+                Some(ref mut x) => &mut x.1,
+                None => {
+                    // This just means no P4 program has been set by the guest.
+                    break;
+                }
+            };
+
+            PacketHandler::process_guest_packet(
+                pkt,
+                &self.data_handles,
+                pl,
+                &self.log,
+            );
+        }
+
+        if push_used {
+            vq.push_used(&mut chain, &mem);
+        }
+    }
+
+    fn handle_q0_req(&self, _vq: &Arc<VirtQueue>) {
+        // This is the queue that the virtio driver in the guest reads from and
+        // we write to. It seems that the correct way to handle a queue
+        // notification on this queue is to not handle it? If we vq.pop_avail to
+        // see what's in the queue, it's always zero data, and the act of doing
+        // a vq.pop_avail seems to drain the queue until it is unusable for
+        // writes, even if we do the corresponding push_used.
+
+        return;
+    }
```

Comment:
The virtio abstraction(s) in propolis should probably grow a way to get/set `USED_F_NO_NOTIFY`.  The kernel side of viona does this, indicating to the guest that we do not want/need notifications for newly available descriptors in the RX ring.

Reply:
In short: Ignoring notifications to vq0 is the right thing to do here.